File: cs2dart_roslyn_worker/Models/ResolvedSymbol.cs
namespace Cs2DartRoslynWorker.Models;

/// <summary>
/// The kind of a resolved symbol.
/// </summary>
public enum SymbolKind
{
    Type,
    Method,
    Field,
    Property,
    Event,
    Local,
    Parameter,
    /// <summary>Sentinel: Roslyn could not bind this reference.</summary>
    Unresolved,
}

/// <summary>
/// A plain-data record representing a fully-resolved Roslyn symbol.
/// No Roslyn types appear in this record.
/// </summary>
public sealed class ResolvedSymbol
{
    /// <summary>Fully-qualified name, e.g. "System.Collections.Generic.List&lt;T&gt;".</summary>
    public string FullyQualifiedName { get; set; } = string.Empty;

    /// <summary>The assembly that defines this symbol, e.g. "System.Collections".</summary>
    public string AssemblyName { get; set; } = string.Empty;

    /// <summary>The kind of symbol.</summary>
    public SymbolKind Kind { get; set; }

    /// <summary>
    /// NuGet package ID when the symbol comes from an external package;
    /// null when the symbol is defined in the same compilation or the BCL.
    /// </summary>
    public string? SourcePackageId { get; set; }

    /// <summary>Source location of the symbol's declaration; null for external symbols.</summary>
    public SourceLocation? SourceLocation { get; set; }

    /// <summary>Compile-time constant value for const symbols; null otherwise.</summary>
    public object? ConstantValue { get; set; }
}

File: lib/src/roslyn_frontend/models/resolved_symbol.dart
import '../../config/models/source_location.dart';

export '../../config/models/source_location.dart';

/// The kind of a resolved symbol.
enum SymbolKind {
  type,
  method,
  field,
  property,
  event,
  local,
  parameter,

  /// Sentinel: Roslyn could not bind this reference.
  unresolved,
}

/// A plain-data record representing a fully-resolved Roslyn symbol.
///
/// No Roslyn types appear in this record. All fields are plain Dart values.
final class ResolvedSymbol {
  /// Fully-qualified name, e.g. "System.Collections.Generic.List<T>".
  final String fullyQualifiedName;

  /// The assembly that defines this symbol, e.g. "System.Collections".
  final String assemblyName;

  /// The kind of symbol.
  final SymbolKind kind;

  /// NuGet package ID when the symbol comes from an external package; null
  /// when the symbol is defined in the same compilation or the BCL.
  final String? sourcePackageId;

  /// Source location of the symbol's declaration; null for external symbols.
  final SourceLocation? sourceLocation;

  /// Compile-time constant value for const symbols; null otherwise.
  final Object? constantValue;

  const ResolvedSymbol({
    required this.fullyQualifiedName,
    required this.assemblyName,
    required this.kind,
    this.sourcePackageId,
    this.sourceLocation,
    this.constantValue,
  });
}
